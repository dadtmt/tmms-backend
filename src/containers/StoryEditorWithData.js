import R from 'ramda'
import { graphql } from 'react-apollo'

import { pageEditorId } from '../config'
import {
  CREATE_CHOICE_MUTATION,
  DELETE_CHOICE_MUTATION,
  CREATE_CROSSROAD_MUTATION,
  CREATE_TEST_MUTATION,
  DELETE_TEST_MUTATION,
  TOGGLE_CROSSROAD_IS_READY,
  UPDATE_CROSSROAD_TEXT
} from '../graphql/mutations'
import { GET_PAGE_EDITOR_QUERY } from '../graphql/queries'
import makeReducer from '../reducers/makeReducer'
import StoryEditor from '../components/StoryEditor'

export const isInteractive = R.cond([
  [
    R.propEq('type', 'dice'),
    R.pipe(R.path(['content', 'master']), R.not)
  ],
  [R.propEq('type', 'characterSheet'), R.F],
  [R.T, R.T]
])

const withCreateChoice = graphql(
  CREATE_CHOICE_MUTATION,
  {
    props: ({ mutate }) => ({
      createChoice: values => mutate({
        variables: {
          newChoice: {
            ...values,
            interactive: isInteractive(values),
            made: false
          }
        }
      })
    })
  }
)

const withDeleteChoice = graphql(
  DELETE_CHOICE_MUTATION,
  {
    props: ({ mutate }) => ({
      deleteChoice: id => mutate({
        variables: {
          choice: {
            id
          }
        }
      })
    })
  }
)

const withCreateCrossroad = graphql(
  CREATE_CROSSROAD_MUTATION,
  {
    props: ({ mutate }) => ({
      createCrossroad: text => mutate({
        variables: {
          newCrossroad: {
            pageEditorId,
            text
          }
        }
      })
    })
  }
)

const withCreateTestDice = graphql(
  CREATE_TEST_MUTATION,
  {
    props: ({ mutate }) => ({
      createTest: values => mutate({
        variables: {
          createTest: {
            ...values,
            made: false
          }
        }
      })
    })
  }
)

const withDeleteTestDice = graphql(
  DELETE_TEST_MUTATION,
  {
    props: ({ mutate }) => ({
      deleteTestDice: id => mutate({
        variables: {
          testDice: {
            id
          }
        }
      })
    })
  }
)


const withToggleCrossroadIsReady = graphql(
  TOGGLE_CROSSROAD_IS_READY,
  {
    props: ({ mutate }) => ({
      toggleIsReady: values => mutate({
        variables: {
          toggleCrossroadReady: {
            ...values
          }
        }
      })
    })
  }
)

const withUpdateCrossroadText = graphql(
  UPDATE_CROSSROAD_TEXT,
  {
    props: ({ mutate }) => ({
      updateCrossroadText: values => mutate({
        variables: {
          updateCrossroadText: {
            ...values
          }
        }
      })
    })
  }
)

const updatePageEditor = update => R.over(
  R.lensPath(['viewer', 'user', 'editors', 'edges']),
  R.over(
    R.lensIndex(0),
    update
  )
)

const updateCrossroad = update => updatePageEditor(
  R.over(
    R.lensPath(['node', 'crossroads', 'edges']),
    R.over(
      R.lensIndex(0),
      update
    )
  )
)

const mutationHandlers = {
  CreateChoice: (state, { createChoice }) => updateCrossroad(
    R.over(
      R.lensPath(['node', 'choices', 'edges']),
      R.append(createChoice.changedEdge)
    )
  )(state),
  CreateCrossroad: (state, { createCrossroad }) => updatePageEditor(
    R.over(
      R.lensPath(['node', 'crossroads', 'edges']),
      R.prepend(createCrossroad.changedEdge)
    )
  )(state),
  CreateSheets: (state, { createSheets }) => updatePageEditor(
    R.over(
      R.lensPath(['node', 'sheets', 'edges']),
      R.append(createSheets.changedEdge)
    )
  )(state),
  CreateTestDice: (state, { createTestDice }) => updateCrossroad(
    R.over(
      R.lensPath(['node', 'testDices', 'edges']),
      R.append(createTestDice.changedEdge)
    )
  )(state),
  DeleteChoice: (state, { deleteChoice }) => updateCrossroad(
    R.over(
      R.lensPath(['node', 'choices', 'edges']),
      R.filter(({ node }) => node.id !== deleteChoice.changedChoice.id)
    )
  )(state),
  DeleteTestDice: (state, { deleteTestDice }) => updateCrossroad(
    R.over(
      R.lensPath(['node', 'testDices', 'edges']),
      R.filter(({ node }) => node.id !== deleteTestDice.changedTestDice.id)
    )
  )(state)
}

const handlers = {
  APOLLO_MUTATION_RESULT: (state, {
    operationName,
    result: { data }
  }) => R.propOr(
    R.identity,
    operationName,
    mutationHandlers
  )(state, data)
}

export const reducer = makeReducer(handlers)

const withData = graphql(
  GET_PAGE_EDITOR_QUERY,
  {
    options: {
      reducer,
      variables: { pageEditorId }
    }
  }
)

export default R.compose(
  withCreateChoice,
  withDeleteChoice,
  withCreateCrossroad,
  withCreateTestDice,
  withDeleteTestDice,
  withData,
  withToggleCrossroadIsReady,
  withUpdateCrossroadText
)(StoryEditor)
