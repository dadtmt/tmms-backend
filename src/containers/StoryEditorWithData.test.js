import React from 'react'
import ReactDOM from 'react-dom'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo'

import StoryEditorWithData, {
  isInteractive,
  reducer
} from './StoryEditorWithData'

const client = new ApolloClient()

describe('isInteractive', () => {
  it('return true by default', () => {
    const values = {}
    expect(isInteractive(values)).toBeTruthy()
  })
  it('return false if type is dice and master true', () => {
    const values = {
      content: {
        master: true
      },
      type: 'dice'
    }
    expect(isInteractive(values)).toBeFalsy()
  })
  it('return true if type is dice and master false', () => {
    const values = {
      content: {
        master: false
      },
      type: 'dice'
    }
    expect(isInteractive(values)).toBeTruthy()
  })
  it('return false if type is characterSheet', () => {
    const values = {
      type: 'characterSheet'
    }
    expect(isInteractive(values)).toBeFalsy()
  })
})

describe('reducer', () => {

  /* eslint-disable sort-keys */
  const state = {
    'viewer': {
      'user': {
        'editors': {
          'edges': [
            {
              'node': {
                'id': 'UGFnZUVkaXRvcjox',
                'crossroads': {
                  'edges': [
                    {
                      'node': {
                        'createdAt': '2017-03-07T11:37:25.000Z',
                        'id': 'Q3Jvc3Nyb2FkOjIx',
                        'isReady': false,
                        'text': {
                          'entityMap': {},
                          'blocks': [
                            {
                              'key': '8ussa',
                              'text': 'crossroad',
                              'type': 'unstyled',
                              'depth': 0,
                              'inlineStyleRanges': [],
                              'entityRanges': [],
                              'data': {}
                            }
                          ]
                        },
                        'choices': {
                          'edges': [
                            {
                              'node': {
                                'id': 'Q2hvaWNlOjY0',
                                'made': false,
                                'text': {
                                  'entityMap': {},
                                  'blocks': [
                                    {
                                      'key': 'f7ujh',
                                      'text': 'Go left',
                                      'type': 'unstyled',
                                      'depth': 0,
                                      'inlineStyleRanges': [],
                                      'entityRanges': [],
                                      'data': {}
                                    }
                                  ]
                                },
                                '__typename': 'Choice'
                              },
                              '__typename': 'ChoiceEdge'
                            }
                          ],
                          '__typename': 'ChoiceConnection'
                        },
                        '__typename': 'Crossroad'
                      },
                      '__typename': 'CrossroadEdge'
                    }
                  ],
                  '__typename': 'CrossroadConnection'
                },
                '__typename': 'PageEditor'
              },
              '__typename': 'UserEditorsEdge'
            }
          ],
          '__typename': 'UserEditorsConnection'
        },
        '__typename': 'User'
      },
      '__typename': 'Viewer'
    }
  }

  /* eslint-enable sort-keys */

  it('return state unchanged on some action', () => {
    expect(reducer(state, { type: 'SOME_ACTION' })).toEqual(state)
  })

  it('handles CREATE_CROSSROAD_MUTATION', () => {

    /* eslint-disable sort-keys */
    const action = {
      'type': 'APOLLO_MUTATION_RESULT',
      'operationName': 'CreateCrossroad',
      'result': {
        'data': {
          'createCrossroad': {
            'changedEdge': {
              'node': {
                'createdAt': '2017-03-07T15:12:18.000Z',
                'id': 'Q3Jvc3Nyb2FkOjIy',
                'isReady': false,
                'text': {
                  'entityMap': {},
                  'blocks': [
                    {
                      'key': '2ovct',
                      'text': 'Here comes the troll...',
                      'type': 'unstyled',
                      'depth': 0,
                      'inlineStyleRanges': [],
                      'entityRanges': [],
                      'data': {}
                    }
                  ]
                },
                'choices': {
                  'edges': [],
                  '__typename': 'ChoiceConnection'
                },
                '__typename': 'Crossroad'
              },
              '__typename': 'CrossroadEdge'
            },
            '__typename': 'CreateCrossroadPayload'
          }
        }
      }
    }

    /* eslint-enable sort-keys */

    expect(reducer(state, action)).toMatchSnapshot()
  })

  it('handles CREATE_CHOICE_MUTATION', () => {

    /* eslint-disable sort-keys */
    const action = {
      'type': 'APOLLO_MUTATION_RESULT',
      'operationName': 'CreateChoice',
      'result': {
        'data': {
          'createChoice': {
            'changedEdge': {
              'node': {
                'id': 'Q2hvaWNlOjU2',
                'made': false,
                'text': {
                  'entityMap': {},
                  'blocks': [
                    {
                      'key': 'nts2',
                      'text': 'This is my choice',
                      'type': 'unstyled',
                      'depth': 0,
                      'inlineStyleRanges': [],
                      'entityRanges': [],
                      'data': {}
                    }
                  ]
                },
                '__typename': 'Choice'
              },
              '__typename': 'ChoiceEdge'
            },
            '__typename': 'CreateChoicePayload'
          }
        }
      }
    }

    /* eslint-enable sort-keys */

    expect(reducer(state, action)).toMatchSnapshot()
  })

  it('handles DELETE_CHOICE_MUTATION', () => {

    /* eslint-disable sort-keys */
    const action = {
      'type': 'APOLLO_MUTATION_RESULT',
      'operationName': 'DeleteChoice',
      'result': {
        'data': {
          'deleteChoice': {
            'changedChoice': {
              'id': 'Q2hvaWNlOjY0',
              '__typename': 'Choice'
            },
            '__typename': 'DeleteChoicePayload'
          }
        }
      }
    }

    /* eslint-enable sort-keys */

    expect(reducer(state, action)).toMatchSnapshot()
  })
})

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    <ApolloProvider client={client} >
      <StoryEditorWithData />
    </ApolloProvider>,
    div
  )
})
