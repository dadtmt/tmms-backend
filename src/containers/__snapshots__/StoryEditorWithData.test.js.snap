exports[`reducer handles CREATE_CHOICE_MUTATION 1`] = `
Object {
  "viewer": Object {
    "__typename": "Viewer",
    "user": Object {
      "__typename": "User",
      "editors": Object {
        "__typename": "UserEditorsConnection",
        "edges": Array [
          Object {
            "__typename": "UserEditorsEdge",
            "node": Object {
              "__typename": "PageEditor",
              "crossroads": Object {
                "__typename": "CrossroadConnection",
                "edges": Array [
                  Object {
                    "__typename": "CrossroadEdge",
                    "node": Object {
                      "__typename": "Crossroad",
                      "choices": Object {
                        "__typename": "ChoiceConnection",
                        "edges": Array [
                          Object {
                            "__typename": "ChoiceEdge",
                            "node": Object {
                              "__typename": "Choice",
                              "id": "Q2hvaWNlOjY0",
                              "made": false,
                              "text": Object {
                                "blocks": Array [
                                  Object {
                                    "data": Object {},
                                    "depth": 0,
                                    "entityRanges": Array [],
                                    "inlineStyleRanges": Array [],
                                    "key": "f7ujh",
                                    "text": "Go left",
                                    "type": "unstyled",
                                  },
                                ],
                                "entityMap": Object {},
                              },
                            },
                          },
                          Object {
                            "__typename": "ChoiceEdge",
                            "node": Object {
                              "__typename": "Choice",
                              "id": "Q2hvaWNlOjU2",
                              "made": false,
                              "text": Object {
                                "blocks": Array [
                                  Object {
                                    "data": Object {},
                                    "depth": 0,
                                    "entityRanges": Array [],
                                    "inlineStyleRanges": Array [],
                                    "key": "nts2",
                                    "text": "This is my choice",
                                    "type": "unstyled",
                                  },
                                ],
                                "entityMap": Object {},
                              },
                            },
                          },
                        ],
                      },
                      "createdAt": "2017-03-07T11:37:25.000Z",
                      "id": "Q3Jvc3Nyb2FkOjIx",
                      "isReady": false,
                      "text": Object {
                        "blocks": Array [
                          Object {
                            "data": Object {},
                            "depth": 0,
                            "entityRanges": Array [],
                            "inlineStyleRanges": Array [],
                            "key": "8ussa",
                            "text": "crossroad",
                            "type": "unstyled",
                          },
                        ],
                        "entityMap": Object {},
                      },
                    },
                  },
                ],
              },
              "id": "UGFnZUVkaXRvcjox",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`reducer handles CREATE_CROSSROAD_MUTATION 1`] = `
Object {
  "viewer": Object {
    "__typename": "Viewer",
    "user": Object {
      "__typename": "User",
      "editors": Object {
        "__typename": "UserEditorsConnection",
        "edges": Array [
          Object {
            "__typename": "UserEditorsEdge",
            "node": Object {
              "__typename": "PageEditor",
              "crossroads": Object {
                "__typename": "CrossroadConnection",
                "edges": Array [
                  Object {
                    "__typename": "CrossroadEdge",
                    "node": Object {
                      "__typename": "Crossroad",
                      "choices": Object {
                        "__typename": "ChoiceConnection",
                        "edges": Array [],
                      },
                      "createdAt": "2017-03-07T15:12:18.000Z",
                      "id": "Q3Jvc3Nyb2FkOjIy",
                      "isReady": false,
                      "text": Object {
                        "blocks": Array [
                          Object {
                            "data": Object {},
                            "depth": 0,
                            "entityRanges": Array [],
                            "inlineStyleRanges": Array [],
                            "key": "2ovct",
                            "text": "Here comes the troll...",
                            "type": "unstyled",
                          },
                        ],
                        "entityMap": Object {},
                      },
                    },
                  },
                  Object {
                    "__typename": "CrossroadEdge",
                    "node": Object {
                      "__typename": "Crossroad",
                      "choices": Object {
                        "__typename": "ChoiceConnection",
                        "edges": Array [
                          Object {
                            "__typename": "ChoiceEdge",
                            "node": Object {
                              "__typename": "Choice",
                              "id": "Q2hvaWNlOjY0",
                              "made": false,
                              "text": Object {
                                "blocks": Array [
                                  Object {
                                    "data": Object {},
                                    "depth": 0,
                                    "entityRanges": Array [],
                                    "inlineStyleRanges": Array [],
                                    "key": "f7ujh",
                                    "text": "Go left",
                                    "type": "unstyled",
                                  },
                                ],
                                "entityMap": Object {},
                              },
                            },
                          },
                        ],
                      },
                      "createdAt": "2017-03-07T11:37:25.000Z",
                      "id": "Q3Jvc3Nyb2FkOjIx",
                      "isReady": false,
                      "text": Object {
                        "blocks": Array [
                          Object {
                            "data": Object {},
                            "depth": 0,
                            "entityRanges": Array [],
                            "inlineStyleRanges": Array [],
                            "key": "8ussa",
                            "text": "crossroad",
                            "type": "unstyled",
                          },
                        ],
                        "entityMap": Object {},
                      },
                    },
                  },
                ],
              },
              "id": "UGFnZUVkaXRvcjox",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`reducer handles DELETE_CHOICE_MUTATION 1`] = `
Object {
  "viewer": Object {
    "__typename": "Viewer",
    "user": Object {
      "__typename": "User",
      "editors": Object {
        "__typename": "UserEditorsConnection",
        "edges": Array [
          Object {
            "__typename": "UserEditorsEdge",
            "node": Object {
              "__typename": "PageEditor",
              "crossroads": Object {
                "__typename": "CrossroadConnection",
                "edges": Array [
                  Object {
                    "__typename": "CrossroadEdge",
                    "node": Object {
                      "__typename": "Crossroad",
                      "choices": Object {
                        "__typename": "ChoiceConnection",
                        "edges": Array [],
                      },
                      "createdAt": "2017-03-07T11:37:25.000Z",
                      "id": "Q3Jvc3Nyb2FkOjIx",
                      "isReady": false,
                      "text": Object {
                        "blocks": Array [
                          Object {
                            "data": Object {},
                            "depth": 0,
                            "entityRanges": Array [],
                            "inlineStyleRanges": Array [],
                            "key": "8ussa",
                            "text": "crossroad",
                            "type": "unstyled",
                          },
                        ],
                        "entityMap": Object {},
                      },
                    },
                  },
                ],
              },
              "id": "UGFnZUVkaXRvcjox",
            },
          },
        ],
      },
    },
  },
}
`;
